util/perwindowlayout/perwindowlayout.lisp

```diff
@@ -0,0 +1,54 @@
+(in-package #:perwindowlayout)
+
+(export '(
+  switch-window-layout
+  enable-per-window-layout
+  disable-per-window-layout
+  ))
+
+(defun get-current-layout (display)
+  (xlib:device-state-locked-group (xlib:get-state display)))
+
+(defun window-focus-changed (window previous-window)
+  (declare (ignore previous-window))
+  (when window
+    (let ((window-layout (getf (xlib:window-plist (window-xwin window)) :keyboard-layout)))
+      (if window-layout
+          (xlib:lock-group *display* :group window-layout)
+          (xlib:lock-group *display* :group 0)))))
+
+(defun group-focus-changed (group previous-group)
+  (let ((previous-window (group-current-window previous-group))
+        (window (group-current-window group)))
+    (window-focus-changed window previous-window)))
+
+(defcommand enable-per-window-layout () ()
+  "Enable layout switching"
+  (xlib::initialize-extensions *display*) ;; we need it because
+  (xlib:enable-xkeyboard *display*) ;; stumpwm opens display before extension definition
+  (add-hook *focus-group-hook* 'group-focus-changed)
+  (add-hook *focus-window-hook* 'window-focus-changed))
+
+(defcommand disable-per-window-layout () ()
+  "Disable layout switching"
+  (remove-hook *focus-window-hook* 'window-focus-changed)
+  (remove-hook *focus-group-hook* 'group-focus-changed))
+
+(defun toggle-window-layout ()
+  (let ((cur-window (current-window)))
+    (when cur-window
+      (let ((current-window-class (window-class cur-window)))
+        (if (string= current-window-class "Emacs")
+            (meta (kbd "C-\\"))
```

Comment:
Definitely, I saw this code to often to get my eyes blurred . So maybe we may go further and use a defparameter, for example
`(defparameter *emacs-toggle-input-method-key* "C-\\")`
What do you think of it?